modules/client/src/internal/utils.ts

```diff
@@ -824,3 +863,763 @@ export function toPluginPermissionOperationType(
+  // return the validity of the proposal
+  return { isValid: causes.length === 0, causes };
+}
+
+export async function validateUpdatePluginProposalActions(
+  actions: DaoAction[],
+  daoAddress: string,
+  pspAddress: string,
+  graphql: IClientGraphQLCore,
+  ipfs: IClientIpfsCore,
+) {
+  let causes: PluginUpdateProposalInValidityCause[][] = [];
+  const classifiedActions = classifyProposalActions(actions);
+  if (isPluginUpdateAction(classifiedActions)) {
+    causes[0] = [];
+    for (const [index, action] of classifiedActions.entries()) {
+      switch (action) {
+        case ProposalActionTypes.GRANT_PLUGIN_UPDATE_PERMISSION:
+          causes[0].concat(
+            await validateGrantUpdatePluginPermissionAction(
+              actions[index],
+              pspAddress,
+              daoAddress,
+              graphql,
+            ),
+          );
+          break;
+        case ProposalActionTypes.REVOKE_PLUGIN_UPGRADE_PERMISSION:
+          causes[0].concat(
+            await validateRevokeUpdatePluginPermissionAction(
+              actions[index],
+              pspAddress,
+              daoAddress,
+              graphql,
+            ),
+          );
+          break;
+        case ProposalActionTypes.APPLY_UPDATE:
+          const resCauses = await validateApplyUpdateFunction(
+            actions[index],
+            daoAddress,
+            graphql,
+            ipfs,
+          );
+          causes[0].concat(
+            resCauses,
+          );
+          break;
+      }
+    }
+    actions = actions.slice(3);
+    if (actions.length !== 0) {
+      const recCauses = await validateUpdatePluginProposalActions(
+        actions,
+        daoAddress,
+        pspAddress,
+        graphql,
+        ipfs,
+      );
+      causes = causes.concat(recCauses.causes);
+    }
+  } else if (isPluginUpdateActionWithRootPermission(classifiedActions)) {
+    causes[0] = [];
+    for (const [index, action] of classifiedActions.entries()) {
+      switch (action) {
+        case ProposalActionTypes.GRANT_PLUGIN_UPDATE_PERMISSION:
+          causes[0].concat(
+            await validateGrantUpdatePluginPermissionAction(
+              actions[index],
+              pspAddress,
+              daoAddress,
+              graphql,
+            ),
+          );
+          break;
+        case ProposalActionTypes.REVOKE_PLUGIN_UPGRADE_PERMISSION:
+          causes[0].concat(
+            await validateRevokeUpdatePluginPermissionAction(
+              actions[index],
+              pspAddress,
+              daoAddress,
+              graphql,
+            ),
+          );
+          break;
+        case ProposalActionTypes.GRANT_ROOT_PERMISSION:
+          causes[0].concat(
+            validateGrantRootPermissionAction(
+              actions[index],
+              daoAddress,
+              pspAddress,
+            ),
+          );
+          break;
+        case ProposalActionTypes.REVOKE_ROOT_PERMISSION:
+          causes[0].concat(
```

Comment:
concat doesn't update the first array but only returns the merged array
```suggestion
          causes[0] = causes[0].concat(
```